src/types.rs

```diff
@@ -1,9 +1,80 @@
 use crate::errors;
-use std::{iter::FromIterator, result};
+use std::{
+    iter::FromIterator,
+    result,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
 
 /// A time tag in OSC message consists of two 32-bit integers where the first one denotes the number of seconds since 1900-01-01 and the second the fractions of a second.
 /// For details on its semantics see http://opensoundcontrol.org/node/3/#timetags
-pub type OscTime = (u32, u32);
+///
+/// # Examples
+///
+/// ```
+/// use rosc::OscTime;
+/// use std::time::SystemTime;
+///
+/// assert_eq!(OscTime::from(SystemTime::UNIX_EPOCH), OscTime::from((2_208_988_800, 0)));
+/// ```
+///
+/// # Conversions between [`std::time::SystemTime`]
+///
+/// The [`From`]/[`Into`] traits are implemented to allow conversions between
+/// [`SystemTime`](std::time::SystemTime) and `OscTime` in both directions. **These conversions are
+/// lossy**, but are tested to have a deviation within 5 nanoseconds when converted back and forth
+/// in either direction.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct OscTime {
+    pub seconds: u32,
+    pub fractional: u32,
+}
+
+impl OscTime {
+    const UNIX_OFFSET: u64 = 2_208_988_800; // From RFC 5905
+    const TWO_POW_32: f64 = 4294967296.0; // Number of bits in a `u32`
+    const NANOS_PER_SECOND: f64 = 1.0e9;
+
+    fn epoch() -> SystemTime {
+        UNIX_EPOCH - Duration::from_secs(OscTime::UNIX_OFFSET)
+    }
+}
+
+impl From<SystemTime> for OscTime {
+    fn from(time: SystemTime) -> OscTime {
+        let duration_since_epoch = time.duration_since(OscTime::epoch()).unwrap();
+        let seconds = duration_since_epoch.as_secs() as u32;
+        let nanos = duration_since_epoch.subsec_nanos() as f64;
+        let fractional = ((nanos * OscTime::TWO_POW_32) / OscTime::NANOS_PER_SECOND).round() as u32;
+        OscTime {
+            seconds,
+            fractional,
+        }
+    }
+}
+
+impl From<OscTime> for SystemTime {
+    fn from(time: OscTime) -> SystemTime {
+        let nanos = (time.fractional as f64) * OscTime::NANOS_PER_SECOND / OscTime::TWO_POW_32;
```

Comment:
It took me some time to wrap my head around the conversion, mainly because I hadn't had to do any serious math for quite some time 😅 

I think that multiplying `time.fractional` with the reciprocal of `TWO_POW_32` and then multiplying by `NANOS_PER_SECOND` is numerically beneficial, since `time.fractional * NANOS_PER_SECOND` can get huge pretty quickly, even risking overflows.  So, what I suggest is to change it to:

```rust
(time.fractional as u64) * OscTime::FRAC_MIN * OscTime::NANOS_PER_SECOND
```

where `OscTime::FRAC_MIN` is `1.0/(u32::MAX as f64)`.   I know that `FRAC_MIN` is not a good name, but I'm open for suggestions.

Reply:
I implemented this and a similar change to remove the division in the conversion going the other direction. I added test causes to check `(u32::MAX, u32::MAX)` and numbers nearby.